File: rpcstream/rpcstream.ts
import { pushable, Pushable } from 'it-pushable'
import { Source, Sink } from 'it-stream-types'
import { RpcAck, RpcStreamPacket } from './rpcstream_pb.js'
import { OpenStreamFunc, PacketStream } from '../srpc/stream.js'
import { PartialMessage } from '@bufbuild/protobuf'

// RpcStreamCaller is the RPC client function to start a RpcStream.
export type RpcStreamCaller = (
  request: AsyncIterable<PartialMessage<RpcStreamPacket>>,
) => AsyncIterable<RpcStreamPacket>

// openRpcStream attempts to open a stream over a RPC call.
// if waitAck is set, waits for the remote to ack the stream before returning.
export async function openRpcStream(
  componentId: string,
  caller: RpcStreamCaller,
  waitAck?: boolean,
): Promise<PacketStream> {
  const packetTx = pushable<PartialMessage<RpcStreamPacket>>({
    objectMode: true,
  })
  const packetRx = caller(packetTx)

  // write the component id
  packetTx.push({
    body: {
      case: 'init',
      value: { componentId },
    },
  })

  // construct packet stream
  const packetIt = packetRx[Symbol.asyncIterator]()

  // wait for ack, if set.
  if (waitAck) {
    const ackPacketIt = await packetIt.next()
    if (ackPacketIt.done) {
      throw new Error(`rpcstream: closed before ack packet`)
    }
    const ackPacket = ackPacketIt.value
    const ackBody = ackPacket?.body
    if (!ackBody || ackBody.case !== 'ack') {
      const msgType = ackBody?.case || 'none'
      throw new Error(`rpcstream: expected ack packet but got ${msgType}`)
    }
    const errStr = ackBody.value?.error
    if (errStr) {
      throw new Error(`rpcstream: remote: ${errStr}`)
    }
  }

  // build & return the data stream
  return new RpcStream(packetTx, packetIt)
}

// buildRpcStreamOpenStream builds a OpenStream func with a RpcStream.
export function buildRpcStreamOpenStream(
  componentId: string,
  caller: RpcStreamCaller,
): OpenStreamFunc {
  return async (): Promise<PacketStream> => {
    return openRpcStream(componentId, caller)
  }
}

// RpcStreamHandler handles an incoming RPC stream.
// implemented by server.handleDuplex.
// return the result of pipe()
export type RpcStreamHandler = (stream: PacketStream) => Promise<void>

// RpcStreamGetter looks up the handler to use for the given Component ID.
// If null is returned, throws an error: "not implemented"
export type RpcStreamGetter = (
  componentId: string,
) => Promise<RpcStreamHandler | null>

// handleRpcStream handles an incoming RPC stream (remote is the initiator).
export async function* handleRpcStream(
  packetRx: AsyncIterator<RpcStreamPacket>,
  getter: RpcStreamGetter,
): AsyncIterable<PartialMessage<RpcStreamPacket>> {
  // read the component id
  const initRpcStreamIt = await packetRx.next()
  if (initRpcStreamIt.done) {
    throw new Error('closed before init received')
  }

  const initRpcStreamPacket = initRpcStreamIt.value

  // ensure we received an init packet
  if (initRpcStreamPacket?.body?.case !== 'init') {
    throw new Error('expected init packet')
  }

  // lookup the server for the component id.
  let handler: RpcStreamHandler | null = null
  let err: Error | undefined
  try {
    handler = await getter(initRpcStreamPacket.body.value.componentId)
  } catch (errAny) {
    err = errAny as Error
    if (!err) {
      err = new Error(`rpc getter failed`)
    } else if (!err.message) {
      err = new Error(`rpc getter failed: ${err}`)
    }
  }

  if (!handler && !err) {
    err = new Error('not implemented')
  }

  yield* [
    {
      body: {
        case: 'ack' as const,
        value: {
          error: err?.message || '',
        },
      },
    },
  ]

  if (err) {
    throw err
  }

  // build the outgoing packet sink & the packet source
  const packetTx: Pushable<RpcStreamPacket> = pushable({ objectMode: true })

  // start the handler
  const rpcStream = new RpcStream(packetTx, packetRx)
  handler!(rpcStream)
    .catch((err) => packetTx.end(err))
    .then(() => packetTx.end())

  // process packets
  for await (const packet of packetTx) {
    yield* [packet]
  }
}

// RpcStream implements the PacketStream on top of a RPC call.
// Note: expects the stream to already have been negotiated.
export class RpcStream implements PacketStream {
  // source is the source for incoming Uint8Array packets.
  public readonly source: AsyncGenerator<Uint8Array>
  // sink is the sink for outgoing Uint8Array packets.
  public readonly sink: Sink<Source<Uint8Array>, Promise<void>>

  // _packetRx receives packets from the remote.
  private readonly _packetRx: AsyncIterator<RpcStreamPacket>
  // _packetTx writes packets to the remote.
  private readonly _packetTx: {
    push: (val: PartialMessage<RpcStreamPacket>) => void
    end: (err?: Error) => void
  }

  // packetTx writes packets to the remote.
  // packetRx receives packets from the remote.
  constructor(
    packetTx: Pushable<PartialMessage<RpcStreamPacket>>,
    packetRx: AsyncIterator<RpcStreamPacket>,
  ) {
    this._packetTx = packetTx
    this._packetRx = packetRx
    this.sink = this._createSink()
    this.source = this._createSource()
  }

  // _createSink initializes the sink field.
  private _createSink(): Sink<Source<Uint8Array>, Promise<void>> {
    return async (source: Source<Uint8Array>) => {
      try {
        for await (const arr of source) {
          this._packetTx.push({
            body: { case: 'data', value: arr },
          })
        }
        this._packetTx.end()
      } catch (err) {
        this._packetTx.end(err as Error)
      }
    }
  }

  // _createSource initializes the source field.
  private _createSource(): AsyncGenerator<Uint8Array> {
    return (async function* (packetRx: AsyncIterator<RpcStreamPacket>) {
      while (true) {
        const msgIt = await packetRx.next()
        if (msgIt.done) {
          return
        }
        const value = msgIt.value
        const body = value?.body
        if (!body) {
          continue
        }
        switch (body.case) {
          case 'ack':
            if (body.value.error?.length) {
              throw new Error(body.value.error)
            }
            break
          case 'data':
            yield body.value
            break
        }
      }
    })(this._packetRx)
  }
}

File: rpcstream/rpcstream.go
package rpcstream

import (
	"context"

	"github.com/aperturerobotics/starpc/srpc"
	"github.com/pkg/errors"
)

// RpcStream implements a RPC call stream over a RPC call. Used to implement
// sub-components which have a different set of services & calls available.
type RpcStream interface {
	srpc.Stream
	Send(*RpcStreamPacket) error
	Recv() (*RpcStreamPacket, error)
}

// RpcStreamGetter returns the Mux for the component ID from the remote.
// Returns a release function to call when done with the Mux.
// Returns nil, nil, nil if not found.
type RpcStreamGetter func(ctx context.Context, componentID string) (srpc.Invoker, func(), error)

// RpcStreamCaller is a function which starts the RpcStream call.
type RpcStreamCaller[T RpcStream] func(ctx context.Context) (T, error)

// OpenRpcStream opens a RPC stream with a remote.
//
// if waitAck is set, waits for acknowledgment from the remote before returning.
func OpenRpcStream[T RpcStream](ctx context.Context, rpcCaller RpcStreamCaller[T], componentID string, waitAck bool) (RpcStream, error) {
	// open the rpc stream
	rpcStream, err := rpcCaller(ctx)
	if err != nil {
		return nil, err
	}

	// write the component id
	err = rpcStream.Send(&RpcStreamPacket{
		Body: &RpcStreamPacket_Init{
			Init: &RpcStreamInit{
				ComponentId: componentID,
			},
		},
	})
	if err != nil {
		_ = rpcStream.Close()
		return nil, err
	}

	// wait for ack
	if waitAck {
		pkt, err := rpcStream.Recv()
		if err == nil {
			switch b := pkt.GetBody().(type) {
			case *RpcStreamPacket_Ack:
				if errStr := b.Ack.GetError(); errStr != "" {
					err = errors.Errorf("remote: %s", errStr)
				}
			default:
				err = errors.New("expected ack packet")
			}
		}
		if err != nil {
			_ = rpcStream.Close()
			return nil, err
		}
	}

	return rpcStream, nil
}

// NewRpcStreamOpenStream constructs an OpenStream function with a RpcStream.
//
// if waitAck is set, OpenStream waits for acknowledgment from the remote.
func NewRpcStreamOpenStream[T RpcStream](rpcCaller RpcStreamCaller[T], componentID string, waitAck bool) srpc.OpenStreamFunc {
	return func(ctx context.Context, msgHandler srpc.PacketDataHandler, closeHandler srpc.CloseHandler) (srpc.PacketWriter, error) {
		// open the stream
		rw, err := OpenRpcStream(ctx, rpcCaller, componentID, waitAck)
		if err != nil {
			return nil, err
		}

		// start the read pump
		go ReadPump(rw, msgHandler, closeHandler)
		// return the writer
		return NewRpcStreamWriter(rw), nil
	}
}

// NewRpcStreamClient constructs a Client which opens streams with a RpcStream.
//
// if waitAck is set, OpenStream waits for acknowledgment from the remote.
func NewRpcStreamClient[T RpcStream](rpcCaller RpcStreamCaller[T], componentID string, waitAck bool) srpc.Client {
	openStream := NewRpcStreamOpenStream(rpcCaller, componentID, waitAck)
	return srpc.NewClient(openStream)
}

// HandleRpcStream handles an incoming RPC stream (remote is the initiator).
func HandleRpcStream(stream RpcStream, getter RpcStreamGetter) error {
	// Read the "init" packet.
	initPkt, err := stream.Recv()
	if err != nil {
		return err
	}
	initInner, ok := initPkt.GetBody().(*RpcStreamPacket_Init)
	if !ok || initInner.Init == nil {
		return ErrUnexpectedPacket
	}

	// lookup the server for this component id
	ctx := stream.Context()
	componentID := initInner.Init.GetComponentId()
	mux, muxRel, err := getter(ctx, componentID)
	if err == nil && mux == nil {
		err = ErrNoServerForComponent
	}
	if mux != nil && muxRel != nil {
		defer muxRel()
	}

	// send ack
	var errStr string
	if err != nil {
		errStr = err.Error()
	}
	sendErr := stream.Send(&RpcStreamPacket{
		Body: &RpcStreamPacket_Ack{
			Ack: &RpcAck{Error: errStr},
		},
	})
	if err != nil {
		return err
	}
	if sendErr != nil {
		return sendErr
	}

	// handle the rpc
	serverRPC := srpc.NewServerRPC(ctx, mux, NewRpcStreamWriter(stream))
	go ReadPump(stream, serverRPC.HandlePacketData, serverRPC.HandleStreamClose)
	return serverRPC.Wait(ctx)
}
